### ISSUE TITLE
Support for generating slimmer packages

### ISSUE THREAD

**Issue by figelwump** [Miserlou/Zappa#1525]
Currently zappa packaging will include all pip packages installed in the virtualenv. Installing zappa in the venv brings in a ton of dependencies. Depending on the app's actual needs, most/all of these don't actually need to be packaged and shipped to lambda. This unnecessarily increases the size of the package which makes zappa deploy/update much slower than it would otherwise.

As an example, for a simple hello world app, the package is over 8MB. The vast majority of this data is unneeded. 

A possible approach here is to have an option to:
- don't package up anything from venv
- use requirements.txt in a way that doesn't slow deploy down

I see #525 and #542 but they don't seem to be resolved yet. Let me know if I'm missing anything!

**Comment by figelwump** [comment]
Maybe zappa-runtime could be split out from the rest of zappa, and only that piece (plus its dependencies) are packaged up? (Of course this only makes sense if the runtime piece has much less dependencies than the rest of it.)

**Comment by scoates** [comment]
Maybe so, but how would you run `zappa` (non-runtime) then?

**Comment by figelwump** [comment]
I'm not sure I understand. Does zappa non-runtime have a dependency on zappa runtime piece?

**Comment by scoates** [comment]
What I mean is that Zappa builds a package by inspecting your local virtualenv. In order to do that, Zappa needs to run. In order to do that, Zappa needs to be available (in the local virtualenv).

**Comment by figelwump** [comment]
I see. I think that's where an option to not inspect the local virtualenv would come in. (The same thing that's being tracked in #525)

So it would look like:
- run zappa package/deploy with a `no-venv` option
- when run with this option, zappa uses requirements.txt to package up the app's dependencies
- also, zappa keeps its own "zappa-runtime-requirements.txt" or whatever and packages the requirements for zappa runtime in the same way

Does this make sense?

**Comment by scoates** [comment]
Sort of makes sense, but has practical problems:
- the next time the user does `pip freeze > requirements.txt`, they'll get the hypothetical zappa packager + dependencies
- if zappa builds a virtualenv from `requirements.txt`, it's entirely possible that a user deploys different versions of packages and dependencies than what they've tested

Those things considered, I don't think that it's a bad idea to have an option to "deploy only requirements.txt", and maybe to split Zappa into packager and runtime.

**Comment by charlax** [comment]
This is especially the case since the virtual env often include dev-only dependencies, such as test libraries (e.g. the `faker` package, which is 6MB). It's kind of a waste to see it included in the deployed zip.

**Comment by figelwump** [comment]
> the next time the user does pip freeze > requirements.txt, they'll get the hypothetical zappa packager + dependencies

True, though the user's not any worse off compared to how it works today. Also to @charlax's point, if you have local dev-only dependencies you don't want to pip freeze those and deploy them to prod in any case.

> if zappa builds a virtualenv from requirements.txt, it's entirely possible that a user deploys different versions of packages and dependencies than what they've tested

Agreed, though isn't that generally the case already? Independent of zappa, your requirements.txt and local dev environment could get out of sync. So it doesn't seem that zappa adds any additional issues there (it's something you already need to be aware of if you use requirements.txt for anything).

It sounds like we're in agreement that it's worth pursuing, which is good news :). What would be the next steps?

**Comment by pcolmer** [comment]
Is there any mileage in considering adding support for Pipfile, which is created by pipenv? The reason for suggesting this is because it explicitly has a separate dev section for packages that are only required for development. Therefore, zappa and associated packages could be listed under that section in the file, and that would keep the "production" section cleaner.

I'm not suggesting that this should replace everything else, but it might work as an optional way forwards to better control what is required just for development and not for deployment? I do realise that adopting pipenv would potentially require a lot of workflow change for developers/users …

**Comment by timworx** [comment]
@pcolmer Came here to see if there was any discussion of exactly that.
Particularly since Pipenv is now the recommended way of managing deps and environments, it would be great to ignore packages that are in dev. 

I have a feeling I'll really wind up solving this with a CI pipeline, but it would be nice to have the option to exclude them.

**Comment by Suor** [comment]
One can easily use `requirements.txt` and `dev-requirements.txt` to split them. Which requires zero extra support from anything. 

I wonder if anyone uses tox to make a fresh virtualenv without all the dev tools.

**Comment by lestephane** [comment]
Just so i understand, should the 'zappa' dependency itself be in
requirements.txt or dev-requirements.txt? Thx

**Comment by mtrovo** [comment]
@lestephane from what I understand now it should be on both, on dev because you need to have access to the CLI tools, and on requirements.txt because Zappa creates a WSGI function_handler that's executed to call your code.

As a first step I think it would be good to abstract the runtime dependency to its own package, later I think it should support something like @figelwump said, a`--no-venv` option that's default for deployments and create a clean venv only with runtime dependencies + "`zappa-runtime`".

**Comment by khamaileon** [comment]
> the next time the user does pip freeze > requirements.txt, they'll get the hypothetical zappa packager + dependencies
@scoates: [pip-tools](https://github.com/jazzband/pip-tools) helps you to handle multiple requirements file (prod / dev)
